pkg/skaffold/build/tag/input_digest.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2019 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tag
+
+import (
+	"context"
+	"crypto/md5"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/docker"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/graph"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+)
+
+type inputDigestTagger struct {
+	cfg docker.Config
+	ag  graph.ArtifactGraph
+}
+
+func NewInputDigestTagger(cfg docker.Config, ag graph.ArtifactGraph) (Tagger, error) {
+	return &inputDigestTagger{
+		cfg: cfg,
+		ag:  ag,
+	}, nil
+}
+
+// this variable is for testing only.
+var getDependenciesForArtifact = graph.DependenciesForArtifact
+
+func (t *inputDigestTagger) GenerateTag(_ string, image *latest.Artifact) (string, error) {
+	var inputs []string
+	ctx := context.Background()
+	srcFies, err := getDependenciesForArtifact(ctx, image, t.cfg, nil)
+	if err != nil {
+		return "", err
+	}
+
+	for _, artifactDep := range t.ag.Dependencies(image) {
+		srcOfDep, err := getDependenciesForArtifact(ctx, artifactDep, t.cfg, nil)
+		if err != nil {
+			return "", err
+		}
+		srcFies = append(srcFies, srcOfDep...)
+	}
+
+	sort.Strings(srcFies)
+	for _, d := range srcFies {
+		h, err := fileHasher(d)
+		if err != nil {
+			if os.IsNotExist(err) {
+				logrus.Tracef("skipping dependency for artifact cache calculation, file not found %s: %s", d, err)
+				continue // Ignore files that don't exist
+			}
+
+			return "", fmt.Errorf("getting hash for %q: %w", d, err)
+		}
+		inputs = append(inputs, h)
+	}
+
+	return encode(inputs)
```

Comment:
I have very limited understanding of how this encoder works, but the `json.Encoder` that's being used takes the entire set of files and returns a single encoding based on that set. is it possible to encode each file individually and then combine the encodings to achieve the same result?

it's worth noting that the elements of `inputs` are the *hashes* of each file and its contents, not the entire file contents themselves. so while there could be thousands of files, this would be a set of thousands of hashes, each of which is only a few bytes.

Reply:
the input array contains only a file paths, I iterate over them in the encode method hashing as you said content and relative path. Maybe the name is misleading.